### PR TITLE
x/mongo/driver: enable parallel zlib compression and improve zstd decompression

### PR DESCRIPTION
**Pull request by charlievieth**
## Summary

x/mongo/driver: enable parallel zlib compression and improve zstd decompression

This commit fixes a bug where zlib compression was serialized across all
goroutines. This occurred because only one shared zlib decompresser was
instantiated for each compression level which had to be locked when used
and thus preventing concurrent compression.

This commit also slightly improves zstd decompression performance by
using a pool of zstd decoders (instantiating a zstd encoded or decoder
is fairly expensive). It also slightly cleans up the logic used to store
and acquire zstd encoders.

```
goos: darwin
goarch: arm64
pkg: go.mongodb.org/mongo-driver/x/mongo/driver
                                    │ base.20.txt  │             new.20.txt              │
                                    │    sec/op    │   sec/op     vs base                │
CompressPayload/CompressorZLib-10     5387.4µ ± 0%   651.1µ ± 1%  -87.91% (p=0.000 n=20)
CompressPayload/CompressorZstd-10      64.56µ ± 1%   64.10µ ± 0%   -0.72% (p=0.000 n=20)
DecompressPayload/CompressorZLib-10    125.7µ ± 1%   123.7µ ± 0%   -1.60% (p=0.000 n=20)
DecompressPayload/CompressorZstd-10    70.13µ ± 1%   45.80µ ± 1%  -34.70% (p=0.000 n=20)
geomean                                235.3µ        124.0µ       -47.31%

                                    │ base.20.txt  │               new.20.txt               │
                                    │     B/s      │      B/s       vs base                 │
CompressPayload/CompressorZLib-10     365.2Mi ± 0%   3021.4Mi ± 1%  +727.41% (p=0.000 n=20)
CompressPayload/CompressorZstd-10     29.76Gi ± 1%    29.97Gi ± 0%    +0.73% (p=0.000 n=20)
DecompressPayload/CompressorZLib-10   15.28Gi ± 1%    15.53Gi ± 0%    +1.63% (p=0.000 n=20)
DecompressPayload/CompressorZstd-10   27.39Gi ± 1%    41.95Gi ± 1%   +53.13% (p=0.000 n=20)
geomean                               8.164Gi         15.49Gi        +89.77%

                                    │ base.20.txt  │               new.20.txt               │
                                    │     B/op     │     B/op      vs base                  │
CompressPayload/CompressorZLib-10     14.02Ki ± 0%   14.00Ki ± 0%   -0.10% (p=0.000 n=20)
CompressPayload/CompressorZstd-10     3.398Ki ± 0%   3.398Ki ± 0%        ~ (p=1.000 n=20) ¹
DecompressPayload/CompressorZLib-10   2.008Mi ± 0%   2.008Mi ± 0%        ~ (p=1.000 n=20) ¹
DecompressPayload/CompressorZstd-10   4.109Mi ± 0%   1.969Mi ± 0%  -52.08% (p=0.000 n=20)
geomean                               142.5Ki        118.5Ki       -16.82%
¹ all samples are equal

                                    │ base.20.txt  │              new.20.txt              │
                                    │  allocs/op   │ allocs/op   vs base                  │
CompressPayload/CompressorZLib-10       1.000 ± 0%   1.000 ± 0%        ~ (p=1.000 n=20) ¹
CompressPayload/CompressorZstd-10       4.000 ± 0%   4.000 ± 0%        ~ (p=1.000 n=20) ¹
DecompressPayload/CompressorZLib-10     26.00 ± 0%   26.00 ± 0%        ~ (p=1.000 n=20) ¹
DecompressPayload/CompressorZstd-10   104.000 ± 0%   1.000 ± 0%  -99.04% (p=0.000 n=20)
geomean                                 10.20        3.193       -68.69%
¹ all samples are equal
```

## Background & Motivation

N/A

### CODE REVIEW

**Comment by charlievieth** [comment]
@matthewdale the big TLDR here is that the current zlib compressor serializes all outbound messages to MongoDB because there is only one shared zlib encoded, which is rightfully protected by a mutex. This represents a serious performance bottleneck for users of this library.

File: x/mongo/driver/compression.go [hunk 1]
```diff
@@ -26,48 +26,70 @@ type CompressionOpts struct {
...
+			return nil, err
+		}
+		enc := &zlibEncoder{writer: writer, level: level}
+		return enc, nil
 	}
-	encoder := &zlibEncoder{writer: writer, buf: new(bytes.Buffer)}
-	zlibEncoders.Store(level, encoder)
+	// The level is invalid so call zlib.NewWriterLever for the error.
+	_, err := zlib.NewWriterLevel(nil, level)
+	return nil, err
```
Comment (matthewdale) [inline x/mongo/driver/compression.go:73]: If the valid levels are ever updated, this could result in unintentionally returning `nil, nil`. It's better to return an explicit error instead.

```suggestion
	// The level is outside the expected range, return an error.
	return nil, fmt.Errorf("invalid zlib compression level: %d", level)
```

File: x/mongo/driver/compression.go [hunk 2]
```diff
@@ -105,8 +127,15 @@ func CompressPayload(in []byte, opts CompressionOpts) ([]byte, error) {
 	}
 }
 
+var zstdReaderPool = sync.Pool{
+	New: func() interface{} {
+		r, _ := zstd.NewReader(nil)
+		return r
```
Comment (matthewdale) [inline x/mongo/driver/compression.go:133]: While we don't expect this to ever return an error with the current implementation, this could cause a difficult-to-diagnose bug if anyone ever changed the `NewReader` params. It's safer to check the error and panic.

```suggestion
		r, err := zstd.NewReader(nil)
		if err != nil {
			panic(err)
		}
		return r
```

File: x/mongo/driver/testdata/compression.go
Comment (matthewdale) [inline x/mongo/driver/testdata/compression.go:1]: Optional: Using a Go file for the data for the compression test is not the most representative payload and a bit confusing for reviewers. Consider using a collection of BSON blobs instead.

**Review by matthewdale — COMMENTED** [review]
Great performance improvements! Looks good pending a few requested maintainability changes.

File: x/mongo/driver/compression.go (hunk 2 above)
Comment (charlievieth) [reply x/mongo/driver/compression.go:133]: I don't think we need this since our tests cover this case and will panic if this ever returns `nil` so any future changes to `zstd.NewReader` will be caught by the tests when we upgrade (I'll add a regression test just in case). 

I'm also pretty adamant that libraries should never panic.

If for some reason we really wanted to check this error then we could use the below:
```go
func getZstdReader() (*zstd.Decoder, error) {
	if r, _ := zstdReaderPool.Get().(*zstd.Decoder); r != nil {
		return r, nil
	}
	return zstd.NewReader(nil) 
}
```

File: x/mongo/driver/testdata/compression.go
Comment (charlievieth) [reply x/mongo/driver/testdata/compression.go:1]: > Optional: Using a Go file for the data for the compression test is not the most representative payload and a bit confusing for reviewers. Consider using a collection of BSON blobs instead.

100% agree, but I needed to do this to make the benchmarks comparable to previous commits. Before this change the benchmarks used the `compression.go` file itself for benchmarks, which was incredibly confusing because any change I made to the file altered the benchmark results.

One option would be to add some representable BSON blob(s) and add additional benchmarks that use them.

File: x/mongo/driver/compression.go [hunk 3]
```diff
@@ -26,48 +26,70 @@ type CompressionOpts struct {
...
-	encoder, err := zstd.NewWriter(nil, zstd.WithEncoderLevel(level))
-	if err != nil {
-		return nil, err
+	if zstd.SpeedFastest <= level && level <= zstd.SpeedBestCompression {
+		return zstdEncoders[level], nil
 	}
-	zstdEncoders.Store(level, encoder)
-	return encoder, nil
+	// The level is invalid so call zstd.NewWriter for the error.
+	return zstd.NewWriter(nil, zstd.WithEncoderLevel(level))
```
[the remark this answers is not in the document]
Comment (charlievieth) [reply x/mongo/driver/compression.go:50]: Sounds good, we test against this and `zstd.WithEncoderLevel` prevents this from happening but I think it's a good change incase there is a ever a regression in the `zstd` lib.

File: x/mongo/driver/compression.go [hunk 4]
```diff
@@ -26,48 +26,70 @@ type CompressionOpts struct {
 	UncompressedSize int32
 }
 
-var zstdEncoders sync.Map // map[zstd.EncoderLevel]*zstd.Encoder
+func zstdNewWriter(lvl zstd.EncoderLevel) *zstd.Encoder {
```
[the remark this answers is not in the document]
Comment (charlievieth) [reply x/mongo/driver/compression.go:29]: I think renaming `zstdNewWriter` to `mustZstdNewWriter` should make it clearer that this will panic and follows the general naming convention of Go functions that will panic.

The issue with putting this in an `init` function is that it the order in which `init` function calls is a bit tricky (though now formalized with go1.21) this can make it dangerous for any other `init` to use any of the code here that relies on the `zstdEncoders`. 

I'm also of the opinion that without a really really strong use-case packages should **never** have an `init` function (at Lyft I banned the use of `init` functions in libraries unless there was an excellent case for it because of the trouble they caused  - ran Go development there).

If we really wanted to doe this without exposing this function we could use this beast, but IMHO it's not 
```go
var zstdEncoders = func() [zstd.SpeedBestCompression + 1]*zstd.Encoder {
	f := func(lvl zstd.EncoderLevel) *zstd.Encoder {
		enc, err := zstd.NewWriter(nil, zstd.WithEncoderLevel(lvl))
		if err != nil {
			panic(err) // this should never happen
		}
		return enc
	}
	return [zstd.SpeedBestCompression + 1]*zstd.Encoder{
		0:                           nil, // zstd.speedNotSet
		zstd.SpeedFastest:           f(zstd.SpeedFastest),
		zstd.SpeedDefault:           f(zstd.SpeedDefault),
		zstd.SpeedBetterCompression: f(zstd.SpeedBetterCompression),
		zstd.SpeedBestCompression:   f(zstd.SpeedBestCompression),
	}
}
```

[the remark this answers is not in the document]
Comment (matthewdale) [reply x/mongo/driver/compression.go:29]: Renaming it to `mustZstdNewWriter` sounds like a great idea!

File: x/mongo/driver/compression.go (hunk 2 above)
Comment (matthewdale) [reply x/mongo/driver/compression.go:133]: If this is covered by a test case, then that sounds reasonable. Consider this resolved.

File: x/mongo/driver/testdata/compression.go
Comment (matthewdale) [reply x/mongo/driver/testdata/compression.go:1]: Wow, I remember that the compression benchmarks used the contents of `compression.go`, but I hadn't considered the effect that has on the stability of the benchmark results That is very confusing.

We can follow up with a change to use BSON blobs as part of [GODRIVER-2948](https://jira.mongodb.org/browse/GODRIVER-2948). Consider this resolved.

**Review by matthewdale — APPROVED** [review]
Looks good 👍

**Review by qingyang-hu — APPROVED** [review]
👍

**Comment by charlievieth** [comment]
Matt, thank you for taking the time to fix this PR up and merge it!
